### PR TITLE
Maker: Fix warnings about undefined component variables

### PR DESCRIPTION
**Pull request by islemaster**
Makes colorLeds a read-only property that turns into a block in droplet.  Globally defines other component variables for linting purposes, to eliminate false warnings about undefined variables (but they're still treated as raw text in block-mode, so they don't break dropdowns).

Before | After
--- | ---
before | after

After discussion with @mrjoshida this PR also includes a change to how we render maker blocks when maker is _disabled_:
image
- Maker blocks are still recognized, but are colored gray.
- Maker blocks are not included in autocomplete.
- Other Maker defines (component names like `buttonL`) are not globally defined, so they'll generate lint warnings.

I've generalized the 'makeDisabledConfig` helper so we can easily do something similar for other conditionally-enabled APIs.

### CODE REVIEW

**Comment by islemaster** [comment]
Hmm... looks like I [legitimately broke](https://cucumber-logs.s3.amazonaws.com/circle/35538/ChromeLatestWin7_droplet_output.html?versionId=QjLU88x6JzHpSmu.nscXZGjsbEwQI2j5) a UI test with this tiny change?  Investigating...

```
failed Open editcode level and write some autocompleted, tooltipped code

    Scenario: Open editcode level and write some autocompleted, tooltipped code

And there is a Tooltipster tooltip with text "radioButton"

Message:

    timed out after 10 seconds (Selenium::WebDriver::Error::TimeOutError)
./step_definitions/droplet_steps.rb:20:in `/^there is a Tooltipster tooltip with text "([^"]*)"$/'
./support/hooks.rb:36:in `call'
./support/hooks.rb:36:in `block in <top (required)>'
features/droplet.feature:27:in `And there is a Tooltipster tooltip with text "radioButton"'
```

File: apps/src/StudioApp.js [hunk 1]
```diff
@@ -1968,22 +1968,12 @@ StudioApp.prototype.handleEditCode_ = function (config) {
...
       delete config.level.codeFunctions[block.func];
     });
-
-    // Remove onRecordEvent from autocomplete, while still recognizing it as a command
-    const block = config.dropletConfig.blocks.find(block => {
-      return makerDropletBlocks.find(makerBlock => makerBlock.func === block.func);
-    });
-    if (block) {
-      block.noAutocomplete = true;
-    }
```
Comment (islemaster) [inline apps/src/StudioApp.js:1952]: Pretty sure this was copypasta - it probably finds the first maker block and disables autocomplete?

File: apps/src/StudioApp.js [hunk 2]
```diff
@@ -1968,22 +1968,12 @@ StudioApp.prototype.handleEditCode_ = function (config) {
     }
   }
 
-  // Remove maker API blocks from palette and autocomplete, unless maker APIs are enabled.
-  // We didn't have access to project.useMakerAPIs() when dropletConfig
-  // was initialized, so include it initially, and conditionally remove it here.
+  // Remove maker API blocks from palette, unless maker APIs are enabled.
```
Comment (islemaster) [inline apps/src/StudioApp.js:1937]: No longer true - we now know whether maker APIs are to be available early enough to do this in `Applab.init`.

File: apps/src/dropletUtils.js [hunk 3]
```diff
@@ -712,6 +721,37 @@ function getParamFromCodeAtIndex(index, methodName, code) {
...
+ * Given a stating config, generates a "disabled" droplet config that has the
+ * following properties:
+ *  - Removes all categories
+ *  - Removes all blocks from categories
+ *  - Turns all blocks gray
+ *  - Removes all blocks from autocomplete
+ *  - Removes all additional defines
+ * @param {DropletConfig} originalConfig
+ */
+module.exports.makeDisabledConfig = function makeDisabledConfig(originalConfig) {
```
Comment (islemaster) [inline apps/src/dropletUtils.js:734]: Note: I'd like to do a follow-up PR that updates dropletUtils to ES6-style exports; I tried doing that as part of this PR but I think it's best kept separate.

File: apps/src/StudioApp.js (hunk 1 above)
Comment (davidsbailey) [reply apps/src/StudioApp.js:1952]: Yep, I think you're right. glad to see this go 🗑

File: apps/src/StudioApp.js [hunk 4]
```diff
@@ -1934,22 +1934,12 @@ StudioApp.prototype.handleEditCode_ = function (config) {
...
   if (!project.useMakerAPIs()) {
-    //// Remove maker blocks from the palette
+    // Remove maker blocks from the palette
```
Comment (davidsbailey) [inline apps/src/StudioApp.js:1939]: this seems like it shouldn't be necessary given that `makeDisabledConfig` "removes all blocks from categories". If that's not true, maybe you can clarify what the difference is?

File: apps/test/unit/dropletUtilsTest.js [hunk 5]
```diff
@@ -385,4 +383,68 @@ describe('getParamFromCodeAtIndex', () => {
...
+    assert(originalConfig.blocks.some(block => !block.noAutocomplete));
+    assert(disabledConfig.blocks.every(block => block.noAutocomplete));
+  });
+
+  it('removes all additional defines', () => {
+    const disabledConfig = dropletUtils.makeDisabledConfig(originalConfig);
+    expect(originalConfig.additionalPredefValues).not.to.be.empty;
+    expect(disabledConfig.additionalPredefValues).to.be.empty;
+  });
+});
```
Comment (davidsbailey) [inline apps/test/unit/dropletUtilsTest.js:448]: Nice testing!

File: apps/src/StudioApp.js (hunk 4 above)
Comment (islemaster) [reply apps/src/StudioApp.js:1939]: This code removes maker blocks from the toolbox configuration, which is separate from the droplet configuration.  The default App Lab toolbox [still contains maker blocks](https://github.com/code-dot-org/code-dot-org/blob/v2017-04-19.1/lib/cdo/shared_constants.rb#L191-L229), and any given level might also.

I haven't checked what happens if the toolbox configuration tries to reference a block that's present in the droplet configuration but not in any category.  If it's omitted without errors (and we consider that by design, not accidental) then I'd be comfortable removing this code too and leaning on the correct droplet config to "audit" the toolbox contents for us.

**Pull request merged**